### DISCUSSION TITLE
Space inside parens () , brackets [] {}

### DISCUSSION THREAD

**Discussion by annalogue-codes** [biomejs/biome#2360] [Ideas]
Dears, thank you for making biome!

I am copying this over from a prettier issue, which reflects a very common coding style.
This has also been discussed on the [prettier issues page](https://github.com/prettier/prettier/issues/1303)

It has become a very common coding style among many frameworks to include spaces inside braces and brackets.
Biome provides the option `javascript.formatter.bracketSpacing` but that does not include braces, i.e. spaces around function parameters and arguments.

#### Function arguments

`function foo( one, two, three ) {`

Instead of

`function foo(one, two, three) {`

#### Arrays

`const array = [ 1, 2, 3, 4 ]`

instead of:

`const array = [1, 2, 3, 4]`

#### Methods

`array.slice( 0, 1 )`

instead of

`array.slice(0, 1)`

### Rationale

 * This coding style strongly increases readability for many developers.
 * It is an actual option in ESLint.
 * It is [used by Wordpress](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/#spacing) and more than Wordpress 50 000 addons.
 * It is part of [Idiomatic.js](https://github.com/rwaldron/idiomatic.js/) by @rwaldron, a fairly popular Javascript style guide that has over 12 000 stars.

**Comment by arendjr** [comment]
I’m not a fan of this option per se, but it does appear there is indeed quite the demand for it. I’m not sure we should prioritize it right now, but I think if someone would be willing to create a PR for this we can accept it.

**Reply by ematipico** [reply]
I would hold on to that, please, and first discuss it.

**Comment by ematipico** [comment]
FYI https://biomejs.dev/formatter/option-philosophy/

**Reply by jonlepage** [reply]
> FYI https://biomejs.dev/formatter/option-philosophy/

hum...
It's kind of like saying to people with disabilities: "Go somewhere else! we don't care about your disability here!" 
**Reading dyslexia** is a fairly common disability, and it is generally well understood, so offering a product or service that is adapted and accessible to people with disabilities enhances your image.
**Not being able to read compact code is not really a matter of preference**, but it **is a real illness**.

**Reply by ematipico** [reply]
Can you provide some context regarding what you just said, the option philosophy and the proposal?

**Reply by jonlepage** [reply]
> Can you provide some context regarding what you just said, the option philosophy and the proposal?

hi,Idon'tunderstandwhatyou'reasking,sorry.Couldyouelaborate?
😉
https://eslint.org/docs/latest/rules/space-in-parens
https://eslint.org/docs/latest/rules/object-curly-spacing
https://eslint.org/docs/latest/rules/array-bracket-spacing

**Reply by jonlepage** [reply]
![image](https://github.com/user-attachments/assets/29391590-b9ac-4edb-bf28-04570ca51af5)

**Reply by ematipico** [reply]
Apologies, but I was looking for something more insightful like discussions or articles regarding the proposal, how a space before parentheses can affect people with disabilities. The two messages gave me nothing, other than just an impression of mocking me.

As you know, Biome formatter is meant to replace Prettier, not Eslint. Also, eslint deprecated stylistic rules.

So, if we are really able to justify really well a new option, we could implement it.

**Reply by annalogue-codes** [reply]
There is plenty of context in the opening post, the discussion in the prettier issue that is linked in the opening post, and in the replies given so far.

Your question comes across like you haven't read the discussion above.

**Reply by annalogue-codes** [reply]
1. It is used in many major frameworks (e.g. Wordpress, jQuery, ... see above).
2. It is more consistent with other rules regarding spaces (see above).
3. It is a profound improvement for a significant number of people with reading dyslexia (and, as the examples under 1. indicate, for a large number of people in general).

How can you just ignore these arguments?

**Reply by annalogue-codes** [reply]
Here is a style guide for dyslexia friendly written English that highlights the importance of spacing:

https://www.bdadyslexia.org.uk/advice/employers/creating-a-dyslexia-friendly-workplace/dyslexia-friendly-style-guide

**Reply by ematipico** [reply]
Thank you, I am going to read the article.

I didn't read the prettier discussion. I just did and it didn't give more information other than the ones you shared here, so you did a good summary.

Btw, please don't assume I am ignoring your arguments. We need to be good at justifying new options, find good arguments, and "because it's used in another tool" isn't one of them, for example. So we need to be good at that! The link might be useful.

**Reply by jonlepage** [reply]
> Apologies, but I was looking for something more insightful like discussions or articles regarding the proposal, how a space before parentheses can affect people with disabilities. The two messages gave me nothing, other than just an impression of mocking me.
> 
> As you know, Biome formatter is meant to replace Prettier, not Eslint. Also, eslint deprecated stylistic rules.
> 
> So, if we are really able to justify really well a new option, we could implement it.

hi, sorry for my inappropriate humor, I was a bit exhausted and my response was probably out of place.
My intention was not to make you feel mocked. If that's the emotion I have given you, I apologize .

The idea is not to criticize but simply to inform you that the choice of including spaces or not is not really about good practice but rather about a medical condition to varying degrees.
English is not my native language, so I wouldn't dare to discuss this topic in a language I don't master well, for fear of expressing myself incorrectly.
However, to provide additional information, here are some articles that discuss the necessity of including spaces to facilitate reading for dyslexic individuals.
Dyslexia has several levels and is sometimes difficult to identify, which can result in a significant decrease in reading efficiency. Generally, a reading task that takes ~1 second to read can take me 5 to 15 seconds and requires a tremendous amount of energy to read text correctly without spaces.

Here some science paper about the sujets:
 
Extra-large letter spacing improves reading in dyslexia
https://pubmed.ncbi.nlm.nih.gov/22665803/

The effects of letter spacing and coloured overlays on reading speed and accuracy in adult dyslexia
https://pubmed.ncbi.nlm.nih.gov/27629067/

Influence of increased letter spacing and font type on the reading ability of dyslexic
https://pubmed.ncbi.nlm.nih.gov/30094714/

french: Dyslexic reading improves when letters are more spaced out.
https://www.ledevoir.com/societe/sante/351667/la-lecture-des-dyslexiques-amelioree-quand-les-lettres-sont-plus-espacees

thanks

**Reply by jonlepage** [reply]
btw, regarding article 2.
**"The effects of letter spacing and coloured overlays on reading speed and accuracy in adult dyslexia"**

People like me often use very colorful syntax to aid in landmarks.
Strangely, this can be very unsettling for those with color blindness or who prefer monochrome, as it often creates too much confusion in their minds.

Without all those colors and spacing, I feel completely lost in my code readings 😅
![image](https://github.com/user-attachments/assets/b92eb036-a981-47d0-ae79-a3f7f66cdca2)

**Comment by ematipico** [comment]
The core team and maintainers discussed the proposal, and **we agree** that we have a strong reasons to **implement** the feature. The main reason is the accessibility concerns. 

Now, for the people that are interested in the proposal, I would ask you to lay down a **concrete** proposal, possibly an [RFC](https://github.com/biomejs/biome/discussions/new?category=rfc), so we can discuss it and come up with a solid plan. 

At the moment we don't have the manpower to implement the feature straightaway, and there are still a lot of things to discuss. Here's some pointers:
- We would prefer to have **one** option to control all the spaces. Dprint has a [`spaceAround`](https://dprint.dev/plugins/typescript/config/) option we could borrow.
- Come up with a name for the option
- Show ALL cases affected by the option. I know it's a lot, but when we will implement the feature, we will need to know where to implement it. For example, the current proposal doesn't cover computed properties. e.g. should the option cover this snippet `{ [someVar]: "" }`? If so, how?

**Reply by Eveeifyeve** [reply]
> The core team and maintainers discussed the proposal, and **we agree** that we have a strong reasons to **implement** the feature. The main reason is the accessibility concerns.
> 
> Now, for the people that are interested in the proposal, I would ask you to lay down a **concrete** proposal, possibly an [RFC](https://github.com/biomejs/biome/discussions/new?category=rfc), so we can discuss it and come up with a solid plan.
> 
> At the moment we don't have the manpower to implement the feature straightaway, and there are still a lot of things to discuss. Here's some pointers:
> 
> * We would prefer to have **one** option to control all the spaces. Dprint has a [`spaceAround`](https://dprint.dev/plugins/typescript/config/) option we could borrow.
> * Come up with a name for the option
> * Show ALL cases affected by the option. I know it's a lot, but when we will implement the feature, we will need to know where to implement it. For example, the current proposal doesn't cover computed properties. e.g. should the option cover this snippet `{ [someVar]: "" }`? If so, how?

Speaking about dprint I am about to look into supporting biome 1.9 for dprint.

**Reply by orballo** [reply]
A while ago, me and @luisherranz, wrote a POC of sorts to see if we could modify Biome to support WordPress coding standard. If I remember correctly, most (if not all) of the changes were related to spacing as it is being discussed here.

Just in case it helps, at least in knowing where to look in the code, here is the PR: https://github.com/orballo/biome/pull/1/files

At the time, we tested the PR against the codebase of [Gutenberg](https://github.com/WordPress/gutenberg), and it seemed to cover all the spacing rules of WP.

**Reply by ematipico** [reply]
Feel free to send a PR @orballo

**Reply by sirreal** [reply]
[Automattic has maintained a fork of prettier (`wp-prettier`) for some time to add the spacing rules expected by the WordPress community.](https://github.com/Automattic/wp-prettier) It may be helpful to look at the changes in that repository, it's a few commits on top of a prettier release commit.

**Reply by Eveeifyeve** [reply]
> A while ago, me and @luisherranz, wrote a POC of sorts to see if we could modify Biome to support WordPress coding standard. If I remember correctly, most (if not all) of the changes were related to spacing as it is being discussed here.
> 
> Just in case it helps, at least in knowing where to look in the code, here is the PR: https://github.com/orballo/biome/pull/1/files
> 
> At the time, we tested the PR against the codebase of [Gutenberg](https://github.com/WordPress/gutenberg), and it seemed to cover all the spacing rules of WP.

I think this should a community thing not a offical support biome in my opinion as wp-prettier is not really used by a lot of users.

**Reply by orballo** [reply]
I didn't mean to imply with my comment that Biome should support WP coding standards.

I shared a POC I worked on because it is related to what is being discussed here, which is to support a different spacing style. It happens that the spacing style is used in the WP coding standards, but that's it.

Also, I'm afraid I won't work on a proper implementation. I just shared the PR with the hope that the person interested in doing so, finds it somewhat helpful.

**Reply by sirreal** [reply]
I'm planning to work with @luisherranz to get a PR ready for this soon, work will start this week.

**Reply by stefanbartos** [reply]
Hey guys, do you have any updates on this? No pressure. Just asking... Thanks

**Reply by luisherranz** [reply]
I'm halfway through, but it's a large PR.

https://github.com/biomejs/biome/compare/main...luisherranz:biome:add-delimiter-spacing-option?expand=1

I don't have much availability, so it'll take me some time to finish, but if anyone wants to help, let me know and we'll coordinate.

**Reply by anton-at-cyolo** [reply]
> I'm halfway through, but it's a large PR.
> 
> https://github.com/biomejs/biome/compare/main...luisherranz:biome:add-delimiter-spacing-option?expand=1
> 
> I don't have much availability, so it'll take me some time to finish, but if anyone wants to help, let me know and we'll coordinate.

Waiting for brackets spacing very much and patiently 🙏

**Reply by jonlepage** [reply]
Claude 4, with Copilot, handled my migration to ESLint 9 and provided several optimizations.  
This setup suits me for now.  
I can continue using ESLint formater, considering it has regained performance thanks to these optimizations!

That was my main reason for looking for an alternative.
ESLint 7 and 8 had very poor performance, required an excessive number of dependencies to install, and configuring them was especially very unpleasant.

The arrival of LLMs changes everything: when I want a rule, I simply ask a agent for it, and it gets found and added automatically in background process!

I'm still interested in testing Biome’s performance and will keep an eye on its development.
